// audit-ci/dist/pnpm-auditer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.audit = exports.report = void 0;
const colors_1 = require("./colors");
const common_1 = require("./common");
const model_1 = __importDefault(require("./model"));
async function runPnpmAudit(config) {
    const { directory, registry, _pnpm, "skip-dev": skipDevelopmentDependencies, } = config;
    const pnpmExec = _pnpm || "pnpm";
    let stdoutBuffer = {};
    function outListener(data) {
        stdoutBuffer = { ...stdoutBuffer, ...data };
    }
    const stderrBuffer = [];
    function errorListener(line) {
        stderrBuffer.push(line);
    }
    const arguments_ = ["audit", "--json"];
    if (registry) {
        console.warn(colors_1.yellow, "PNPM audit does not support the registry flag yet.");
    }
    if (skipDevelopmentDependencies) {
        arguments_.push("--prod");
    }
    const options = { cwd: directory };
    await (0, common_1.runProgram)(pnpmExec, arguments_, options, outListener, errorListener);
    if (stderrBuffer.length > 0) {
        throw new Error(`Invocation of pnpm audit failed:\n${stderrBuffer.join("\n")}`);
    }
    return stdoutBuffer;
}
function printReport(parsedOutput, levels, reportType, outputFormat) {
    const printReportObject = (text, object) => {
        if (outputFormat === "text") {
            console.log(colors_1.blue, text);
        }
        console.log(JSON.stringify(object, undefined, 2));
    };
    switch (reportType) {
        case "full":
            printReportObject("PNPM audit report JSON:", parsedOutput);
            break;
        case "important": {
            const { advisories, metadata } = parsedOutput;
            const relevantAdvisoryLevels = Object.keys(advisories).filter((advisory) => levels[advisories[advisory].severity]);
            const relevantAdvisories = {};
            for (const advisory of relevantAdvisoryLevels) {
                relevantAdvisories[advisory] = advisories[advisory];
            }
            const keyFindings = {
                advisories: relevantAdvisories,
                metadata: metadata,
            };
            printReportObject("PNPM audit report results:", keyFindings);
            break;
        }
        case "summary":
            printReportObject("PNPM audit report summary:", parsedOutput.metadata);
            break;
        default:
            throw new Error(`Invalid report type: ${reportType}. Should be \`['important', 'full', 'summary']\`.`);
    }
}
function report(parsedOutput, config, reporter) {
    const { levels, "report-type": reportType, "output-format": outputFormat, } = config;
    printReport(parsedOutput, levels, reportType, outputFormat);
    const model = new model_1.default(config);
    const summary = model.load(parsedOutput);
    return reporter(summary, config, parsedOutput);
}
exports.report = report;
/**
 * Audit your PNPM project!
 *
 * @returns Returns the audit report summary on resolve, `Error` on rejection.
 */
async function audit(config, reporter = common_1.reportAudit) {
    const parsedOutput = await runPnpmAudit(config);
    if ("error" in parsedOutput) {
        const { code, summary } = parsedOutput.error;
        throw new Error(`code ${code}: ${summary}`);
    }
    return report(parsedOutput, config, reporter);
}
exports.audit = audit;
